include/pybind11/pytypes.h

```diff
@@ -380,20 +383,49 @@ class PYBIND11_EXPORT_EXCEPTION error_already_set : public std::runtime_error {
 public:
     /// Constructs a new exception from the current Python error indicator, if any.  The current
     /// Python error indicator will be cleared.
-    error_already_set() : std::runtime_error(detail::error_string()) {
+    error_already_set() : std::runtime_error("") {
         PyErr_Fetch(&m_type.ptr(), &m_value.ptr(), &m_trace.ptr());
+        if (m_type) {
+            PyErr_NormalizeException(&m_type.ptr(), &m_value.ptr(), &m_trace.ptr());
+            /*if (m_trace) {
+                PyException_SetTraceback(m_value.ptr(), m_trace.ptr());
+            }*/
+        }
     }
 
     error_already_set(const error_already_set &) = default;
     error_already_set(error_already_set &&) = default;
 
     inline ~error_already_set() override;
 
+    const char *what() const noexcept override {
+        if (m_lazy_what.empty()) {
+            try {
+                m_lazy_what = detail::error_string(m_type.ptr(), m_value.ptr(), m_trace.ptr());
+            } catch (const std::exception &e) {
+                m_lazy_what
+                    = std::string(
+                          "Unknown internal error occurred while constructing error_string: ")
+                      + e.what();
+                return m_lazy_what.c_str();
+            } catch (...) {
+                m_lazy_what = "Unknown internal error occurred";
+                return m_lazy_what.c_str();
+            }
+        }
+        assert(!m_lazy_what.empty());
```

Comment:
I think having this fail in production (not just in debug mode) is critical; worth much more than the very minor runtime and binary size savings.

```
if (m_lazy_what.empty()) pybind11_fail("Critical internal error building m_lazy_what.");
```

Reply:
As a reminder, what is noexcept so do we propagate the exception? The whole point here is to catch any errors that error_string() might throw.